### PR TITLE
GUI receives JSON shooting coordinates

### PR DESCRIPTION
**Pull request by AdrianHarenczyk**
+ GUI can receive coordinates and fieldstate 
+ according to that, it changes the way enemy board looks
+ data is mocked and will be replaced with actual server message
+ add class Field which for now represents information about field from server
+ add class FieldState which represents state of that field
+ add util class to process GSON conversion.

### CODE REVIEW

File: client/src/main/java/gui/data/Field.java [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+import java.util.Objects;
+
+public class Field {
+    private Map<Coordinates, FieldState> coordinatesState;
+
+    private Field(Map<Coordinates, FieldState> coordinatesState) {
+        this.coordinatesState = coordinatesState;
+    }
+
+    public static Field create(Map<Coordinates, FieldState> coordinatesState) {
```
Comment (RafalJDev) [inline client/src/main/java/gui/data/Field.java]: ofCoordinateState

File: client/src/main/java/gui/printers/FieldPrinter.java [hunk 2]
```diff
@@ -1,15 +1,23 @@
...
 import javafx.scene.control.Button;
 import javafx.scene.layout.GridPane;
+import model.Coordinates;
 
 /**
  * This class is made to help with printing fields as buttons on enemy board.
  */
 public class FieldPrinter {
 
+    private static final String WATER_STYLE = "water-button";
```
Comment (RafalJDev) [inline client/src/main/java/gui/printers/FieldPrinter.java]: 6 fields ??? Wrap it

No line separators, for common fields

File: client/src/main/java/gui/printers/FieldPrinter.java [hunk 3]
```diff
@@ -23,40 +31,83 @@ public static void insertFields(GridPane printingBoard) {
...
 
     private static void addStyle(Button field) {
-        field.getStyleClass().add("water-button");
+        field.getStyleClass().add(WATER_STYLE);
     }
 
-    private static void addMissOnClick(Button field) {
+    private static void addShotOnAction(Button field, int widthCoordinates, int heightCoordinates) {
         ClientAppRunner appInstance = ClientAppRunner.getInstance();
         ServerConnection serverConnection = appInstance.getServerConnection();
```
Comment (RafalJDev) [inline client/src/main/java/gui/printers/FieldPrinter.java:65]: Separe line

File: client/src/main/java/gui/printers/FieldPrinter.java [hunk 4]
```diff
@@ -23,40 +31,83 @@ public static void insertFields(GridPane printingBoard) {
...
-        for (int widthPosition = 0; widthPosition < numberOfRows; widthPosition++) {
-            for (int heightPosition = 0; heightPosition < numberOfColumns; heightPosition++) {
-                Button fieldToClick = createField();
-                printingBoard.add(fieldToClick, widthPosition, heightPosition);
-            }
+    private static void changeColorFromState(FieldState state, Button button) {
+        if (null == state) {
+            return;
         }
+        switch (state) {
```
Comment (RafalJDev) [inline client/src/main/java/gui/printers/FieldPrinter.java]: Strategia ? Design Pattern

File: client/src/main/java/gui/printers/FieldPrinter.java [hunk 5]
```diff
@@ -23,40 +31,83 @@ public static void insertFields(GridPane printingBoard) {
...
+    private static void addDamagedMark(Button button) {
+        ObservableList<String> styles = button.getStyleClass();
+        styles.removeAll(WATER_STYLE);
+        styles.add(DAMAGED_STYLE);
+    }
+
+    private static void addDestroyedMark(Button button) {
+        ObservableList<String> styles = button.getStyleClass();
+        styles.removeAll(WATER_STYLE);
+        styles.add(DESTROYED_STYLE);
```
Comment (RafalJDev) [inline client/src/main/java/gui/printers/FieldPrinter.java]: No matter what is done in class 100 LoC is screaming, I am not SRP. (ofcourse there may be exceptions).

LoC: Line of Code

Here: Strategy pattern will help

File: client/src/main/java/gui/recievers/ShotResult.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+import gui.data.Field;
+import gui.data.FieldState;
+import gui.utility.JSONConverter;
+import model.Coordinates;
+
+import java.lang.reflect.Type;
+import java.util.Map;
+
+public class ShotResult {
+    private static final Type TYPE =
```
Comment (RafalJDev) [inline client/src/main/java/gui/recievers/ShotResult.java]: Maybe method to JsonHelper, or sth like that

**Review by RafalJDev — APPROVED** [review]
Approve, but you should consider applying my advices :+1:

File: client/src/main/java/gui/controllers/PlacementSceneController.java [hunk 7]
```diff
@@ -4,7 +4,7 @@
 import connection.ServerConnection;
 import gui.instance.ClientAppRunner;
 import gui.printers.FleetView;
-import gui.generator.RandomFleet;
+import gui.recievers.RandomFleet;
```
Comment (KrzysztofDzioba) [inline client/src/main/java/gui/controllers/PlacementSceneController.java]: typo

File: client/src/main/java/gui/data/Field.java [hunk 8]
```diff
@@ -0,0 +1,47 @@
+package gui.data;
+
+import model.Coordinates;
+
+import java.util.Map;
+import java.util.Objects;
+
+public class Field {
+    private Map<Coordinates, FieldState> coordinatesState;
```
Comment (KrzysztofDzioba) [inline client/src/main/java/gui/data/Field.java]: One field on board can have different coordinates with different fieldStates?

File: client/src/main/java/gui/printers/FieldPrinter.java [hunk 9]
```diff
@@ -23,40 +31,83 @@ public static void insertFields(GridPane printingBoard) {
         addToEachSpace(numberOfRows, numberOfColumns, printingBoard);
     }
 
-    private static Button createField() {
+    private static void addToEachSpace(int numberOfRows, int numberOfColumns, GridPane printingBoard) {
+        for (int widthPosition = 0; widthPosition < numberOfRows; widthPosition++) {
+            for (int heightPosition = 0; heightPosition < numberOfColumns; heightPosition++) {
```
Comment (KrzysztofDzioba) [inline client/src/main/java/gui/printers/FieldPrinter.java:35]: IntStreams would be great here!

File: client/src/main/java/gui/recievers/RandomFleet.java [hunk 10]
```diff
@@ -1,12 +1,12 @@
...
 
 
-import com.google.gson.Gson;
 import gui.printers.FleetView;
 import gui.printers.ShipView;
+import gui.utility.JSONConverter;
 
 /**
- * GUI-side fleet random generator
+ * GUI-side fleet random recievers
```
Comment (KrzysztofDzioba) [inline client/src/main/java/gui/recievers/RandomFleet.java]: typo

File: client/src/main/java/gui/controllers/PlacementSceneController.java (hunk 7 above)
Comment (AdrianHarenczyk) [reply client/src/main/java/gui/controllers/PlacementSceneController.java]: Will change

File: client/src/main/java/gui/printers/FieldPrinter.java (hunk 4 above)
Comment (AdrianHarenczyk) [reply client/src/main/java/gui/printers/FieldPrinter.java]: Will do in the future. Tracking in #35

File: client/src/main/java/gui/utility/JSONConverter.java [hunk 11]
```diff
@@ -0,0 +1,16 @@
...
+import java.lang.reflect.Type;
+
+public class JSONConverter {
+    private static Gson convertHandler = new Gson();
+
+    private JSONConverter() {
+    }
+
+    public static <T> T convertToClass(String input, Type type) {
+        return convertHandler.fromJson(input, type);
```
Comment (KrzysztofDzioba) [inline client/src/main/java/gui/utility/JSONConverter.java:14]: can be changed to "new Gson().fromJson(input, type);" so the static field would be removed. But it is your decision to create new object every time and remove static field or keep it like it is

**Review by KrzysztofDzioba — APPROVED** [review]
Looks almost perfect. Small suggestions provided

File: client/src/main/java/gui/utility/JSONConverter.java (hunk 11 above)
Comment (AdrianHarenczyk) [reply client/src/main/java/gui/utility/JSONConverter.java:14]: I would like to leave it like it is. Thanks for the great advice. :)

File: client/src/main/java/gui/recievers/RandomFleet.java (hunk 10 above)
Comment (AdrianHarenczyk) [reply client/src/main/java/gui/recievers/RandomFleet.java]: Thanks for spotting. Didn't noticed it. :+1:

File: client/src/main/java/gui/data/Field.java (hunk 8 above)
Comment (AdrianHarenczyk) [reply client/src/main/java/gui/data/Field.java]: Will change name.

File: client/src/main/java/gui/recievers/ShotResult.java (hunk 6 above)
Comment (AdrianHarenczyk) [reply client/src/main/java/gui/recievers/ShotResult.java]: Change to private method.

File: client/src/main/java/gui/printers/FieldPrinter.java (hunk 5 above)
Comment (AdrianHarenczyk) [reply client/src/main/java/gui/printers/FieldPrinter.java]: Will do in the future. Tracking in #35

File: client/src/main/java/gui/printers/FieldPrinter.java (hunk 9 above)
Comment (AdrianHarenczyk) [reply client/src/main/java/gui/printers/FieldPrinter.java:35]: Thanks it's great idea for future refactor. :+1:

File: client/src/main/java/gui/printers/FieldPrinter.java (hunk 2 above)
Comment (AdrianHarenczyk) [reply client/src/main/java/gui/printers/FieldPrinter.java]: Wraped in enum.

File: client/src/main/java/gui/data/Field.java (hunk 1 above)
Comment (AdrianHarenczyk) [reply client/src/main/java/gui/data/Field.java]: Will leave it like it is. Too long name of factory method in my opinion.

**Comment by AdrianHarenczyk** [comment]
I've included most of changes which were suggested. For now code looks good enough for us to merge. Thank you for spotting some issues.

**Pull request merged**